### DISCUSSION TITLE
Risor to Go interop

### DISCUSSION THREAD

**Discussion by JayJamieson** [deepnoodle-ai/risor#74] [Ideas]
Are there any examples of being able to register Go code in the risor run time to be able to call into Go from risor?

I've been experimenting with v8go for scripting Go from v8 but encountered performance issues. This project looks promising as a possible better alternative.

**Comment by myzie** [comment]
Hi @JayJamieson, this is the example I have on-hand right now:

https://github.com/risor-io/risor/blob/main/examples/go/struct/main.go

The Go service is available in the Risor script, and the script runs this against it:

```go
svc.SetName("My Service")
svc.Start()
svc.PrintState()
print("metrics:", svc.GetMetrics())
print("start count:", svc.GetState()["Counts"]["Start"])
```

**Reply by JayJamieson** [reply]
Ah, Appears I am just blind and stupid when it comes to realizing what Proxy means in the context of code interop. Cheers, I'll give this a go.

**Reply by myzie** [reply]
Naming is hard. It probably wouldn't be obvious to me either, had I not written the code!

**Reply by JayJamieson** [reply]
Are there examples of invoking a risor function from go passing input from Go side to risor?

risor code:
```
func hello(name) { 
    print("Hello" + name)
}
```

Psuedo code of how i imagine it could be called.
```
// ... setup risor code in vm
code := os.ReadFile("hello.risor")
risorVm := vm.New(code)

result, err := risorVm.Invoke("hello", risor.StringParam("world"))
```

**Reply by myzie** [reply]
There's not a built-in way to invoke a function in this way, at this moment. 

For the time being, you could do something like auto-append a line of code that invokes the function as the last line of the code. The return value of the function would be the risor Eval result value.

```go
code := whatever
code += "\n"
code += fmt.Sprintf("hello(\"%s\")", "world")
result, err := risor.Eval(ctx, code, opts...)
```

I can definitely imagine creating a helper to make doing this more natural. What do you think?

**Reply by JayJamieson** [reply]
My initial try was just appending it at the end of the code being eval'd. It works well enough.

Happy to wait for a helper to make the ergonomics of doing it nicer and possibly something similar to the psuedo code I shared. 

Appreciate the discussions so far, if you have some sort of tip jar i would be more than happy to give a few bucks where I can.

**Reply by myzie** [reply]
Thanks dude, I appreciate it. Right now getting feedback like this about what is missing and what could be easier is super valuable, for what it's worth. Thanks for giving it a try even though it has some rough edges :-)
